### PR TITLE
feat(config): add Git::Configuring mixin, parsers, and ConfigEntryInfo

### PR DESCRIPTION
**Pull request by jcouball**
# PR 1 — Config foundation: Git::ConfigEntryInfo, Git::Parsers::ConfigEntry, Git::Configuring

## Summary

Adds the three foundational types for the `git config` redesign (see
`redesign/config_design.rb`). This PR is purely additive — nothing is mixed into
any existing class, so there is zero public API change.

### New files

| File | Purpose |
| --- | --- |
| `lib/git/config_entry_info.rb` | `Git::ConfigEntryInfo` — an immutable `Data` value object representing a single config entry with `scope`, `origin`, `key`, `value`, plus `#section`, `#subsection`, `#variable` accessors |
| `lib/git/parsers/config_entry.rb` | `Git::Parsers::ConfigEntry` — four `module_function` parsers that turn raw `git config` stdout into arrays/single entries of `ConfigEntryInfo`: `parse_get`, `parse_get_all`, `parse_list`, `parse_urlmatch` |
| `lib/git/configuring.rb` | `Git::Configuring` mixin — 13 public `config_*` read/write methods, each delegating to the appropriate `Git::Commands::ConfigOptionSyntax::*` class; includes `Private.assert_valid_opts!` and abstract `#execution_context` / `#assert_valid_scope!` |

### Bug fix in existing command

`lib/git/commands/config_option_syntax/get_urlmatch.rb`: removed `flag_option :show_origin`
because git has never supported `--show-origin` with `--get-urlmatch` (git exits 129 with
"error: --show-origin is only applicable to --get, --get-all, --get-regexp, and --list").
`parse_urlmatch` handles the resulting 2-field-per-entry format (`scope\0key\nvalue\0`)
and sets `origin: nil` on every returned `ConfigEntryInfo`.

### Wire-up in `lib/git.rb`

Added `require` lines for all three new files so they are available when the gem loads.

## Design decisions

- **`parse_urlmatch` is separate from `parse_list`**: the `--get-urlmatch` format with
  `--show-scope --null` (but no `--show-origin`) is structurally different (2 tokens
  instead of 3), so a dedicated parser method keeps the implementation simple and
  the format difference explicit.
- **`CONFIG_*_ALLOWED_OPTS` constants are module-level**: they are `@api private` and
  `private_constant`-protected; they exist at module scope (not method scope) to avoid
  re-allocating frozen arrays on every call.
- **`Private` sub-module**: `assert_valid_opts!` lives in `Git::Configuring::Private` to
  avoid polluting the mixin's public or instance-private namespace.

## Test coverage

**Unit tests** (80 examples):
- `spec/unit/git/config_entry_info_spec.rb` — all three accessor methods across all
  key-shape edge cases (no dot, one dot, multi-dot, subsection with embedded dots)
- `spec/unit/git/parsers/config_entry_spec.rb` — 100% line and branch coverage across
  all four parser methods, including empty-output, single-entry, and multi-entry cases
- `spec/unit/git/configuring_spec.rb` — every public method stubbed at the command
  layer; tests delegation, option merging, unknown-option rejection, `assert_valid_scope!`
  invocation, and abstract method contracts

**Integration tests** (15 examples):
- `spec/integration/git/parsers/config_entry_spec.rb` — each parser method called
  against real `git config` output inside a test repository, validating that the format
  assumptions in unit fixtures match actual git behaviour
- `spec/integration/git/commands/config_option_syntax/get_urlmatch_spec.rb` — existing
  file extended with a `show_scope: true` test; `show_origin: true` test removed (git
  rejects it for `--get-urlmatch`)

## Checklist

- [x] I reviewed and applied the project's [AI Policy](../AI_POLICY.md). I understand
  and verify any AI-assisted changes included in this PR and ensured they meet
  quality, security, and licensing standards.
- [x] I ran `bundle exec rake` locally on this branch and it passed (see
  [Local development setup](../CONTRIBUTING.md#local-development-setup)).
- [x] Tests added/updated as needed.
- [x] Documentation updated where applicable (README and/or YARD).

### CODE REVIEW

File: lib/git/configuring.rb [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+  #   private_class_method :assert_valid_scope!
+  #
+  # @api public
+  #
+  module Configuring # rubocop:disable Metrics/ModuleLength
+    # @!group Read Operations
+
+    # @api private
+    CONFIG_GET_ALLOWED_OPTS = %i[global system local worktree file f blob includes no_includes type default].freeze
+    private_constant :CONFIG_GET_ALLOWED_OPTS
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/configuring.rb:49]: What? No! Keep the default value. Here is what `git config --get` with default actually outputs:

When the key is found:

```
[james] ruby-git $ git config --get --default asdf --show-scope --show-origin user.name
global  file:/Users/james/.gitconfig    James Couball
```

When the key is not found:

```
[james] ruby-git $ git config --get --default asdf --show-scope --show-origin user.bogus
command command line:   asdf
```

File: lib/git/configuring.rb [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+    #
+    # @option options [Boolean, nil] :includes (nil) follow include directives
+    #
+    # @option options [Boolean, nil] :no_includes (nil) suppress include directives
+    #
+    # @option options [String, nil] :type (nil) enforce a type constraint on the value
+    #
+    # @option options [String, nil] :default (nil) value to return when the key is missing
+    #
+    # @return [Git::ConfigEntryInfo, nil] the matching entry, or `nil` when not found
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/configuring.rb:85]: `--default` should be supported

File: lib/git/config_entry_info.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    #
+    # Returns an empty string when the key has zero or one dot (no subsection).
+    #
+    # @example Subsection component of a dotted key
+    #   entry.subsection # => "origin"
+    #
+    # @return [String] the subsection name, or an empty string when there is no subsection
+    #
+    def subsection = first_dot && first_dot != last_dot ? key[(first_dot + 1)...last_dot] : ''
+
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/config_entry_info.rb:80]: This is fine. We can optimize later.
